Document the use of mandatory macro parameters and add named macro tokens

bernt-matthias opened the pull request:
learned while working on https://github.com/galaxyproject/galaxy/pull/13152

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
-macros through the use of specifying ``token_xyz`` attributes on the macro
+macros through the use of specifying 
+
+- required parameters as comma separated list of parameter names using the
```
```suggestion
- required parameters as comma-separated list of parameter names using the
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
+- required parameters as comma separated list of parameter names using the
+``tokens`` attribute of the macro, e.g. ``tokens="foo,bar"`` and then using
+@FOO@ and @BAR@ anywhere in the macro 
+- optional parameters as ``token_xyz`` attributes on the macro 
```
```suggestion
- optional parameters as ``token_xyz="default_value"`` attributes of the ``xml`` 
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
+``tokens`` attribute of the macro, e.g. ``tokens="foo,bar"`` and then using
+@FOO@ and @BAR@ anywhere in the macro 
+- optional parameters as ``token_xyz`` attributes on the macro 
 definition, and then using ``@XYZ@`` anywhere within the XML.
```
```suggestion
  element, and then using ``@XYZ@`` anywhere inside the macro definition.
```

mvdbeek commented on docs/_writing_macros.rst:
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
+          </param>
+        </when>
+        <when value="b">
+          <param ... />
```
I think it helps to have this be somewhat valid XML.

mvdbeek commented on docs/_writing_macros.rst:
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token` tag with the same name.
+Token tags need to be direct children of the ``expand`` element.
+
```
Can we have the example be something like:
Given the following macro:
...
and expanding the macro in the following way:
...
we get the following expanded definition:
...
?

mvdbeek commented on docs/_writing_macros.rst:
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
+
+.. code-block:: xml
+
+        <expand macro="named_yields_example">
```
The indentiation looks weird to me, why not start on the left side ?

mvdbeek commented on docs/_writing_macros.rst:
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
+              </param>
+            </when>
+          </token>
+          <options from_data_table="tophat2_indexes">
+            <filter type="sort_by" column="2"/>
+            <validator type="no_options" message="No genomes are available for the selected input dataset"/>
+          </options>
```
```suggestion
```
This is unrelated to the macro example, right ?

bernt-matthias replied:
The unnamed yield is replaced by this.

mvdbeek replied:
Oh, I have not understood this. Why is that ? How do I read this as a tool author ?

bernt-matthias replied:
Essentially as before: The state was that the content of the expand tag is inserted at the place(s) of the yield tags in the macro. 

In order to keep backward compatibility I implemented it such still all content of the expand minus the token tags and their content is used for the replacement.

mvdbeek replied:
Ah, I see, yeah, that makes sense.

mvdbeek commented on docs/_writing_macros.rst:
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
 
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token` tag with the same name.
+Token tags need to be direct children of the ``expand`` element.
```
```suggestion
Token tags need to be direct children of the ``expand`` element.
This is useful when you would like to compose options of a select list, as shown in the example below.
```

bernt-matthias commented:
I hope that I have addressed all comment. 

Just had the idea that one could allow content to `yield` tags and interpret it as default if the corresponding token is not given. 
Seems that writing docs is not only annoying but may result in better ideas and helps to reason about code.

bernt-matthias commented:
Hey @nsoranzo there were quite a few commits after your approval. Are you still OK with merging?

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+   </expand>
+
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token`` tag with the same name.
```
```suggestion
``<yield name="xyz"/>`` it is replaced by the content of the ``token`` element with the same name.
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token`` tag with the same name.
+Token tags need to be direct children of the ``expand`` element. This is useful if different
```
```suggestion
Token elements need to be direct children of the ``expand`` element. This is useful if different
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+we get the following expanded definition:
+
+.. code-block:: xml
+   :emphasize-lines: 6,9-14,17-24
```
```suggestion
   :emphasize-lines: 6,10-13,17-24
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+   </xml>
+
+Named yields are replaced in the order of the tokens defined in the ``expand``
+tag. Unamed yields are replaced after all named tokens have been replaced (by
```
```suggestion
tag. Unnamed yields are replaced after all named tokens have been replaced (by
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+the non-token child elements of the expand tag). If there are named
+yields that have no corresponding token, then they are treated like unnamed
+yields. Note that, unnamed and named tokens can be used multiple times in a
+macro, then each occurence is preplaced by the corresponding content definedin
```
```suggestion
macro, then each occurrence is replaced by the corresponding content defined in
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
+tag. Unamed yields are replaced after all named tokens have been replaced (by
+the non-token child elements of the expand tag). If there are named
+yields that have no corresponding token, then they are treated like unnamed
+yields. Note that, unnamed and named tokens can be used multiple times in a
```
```suggestion
yields. Note that unnamed and named tokens can be used multiple times in a
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
 Parameterizing XML Macros (with tokens)
 -----------------------------------------
 
 In addition to using ``yield`` blocks, there is another way to parameterize
```
```suggestion
In addition to using ``yield`` blocks, there is another way to parametrize
```

nsoranzo commented on docs/_writing_macros.rst:
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
-    </inputs>
+   <macros>
+     <xml name="color" tokens="attr,attr_value" token_quote="__" token_label="label">
+         <param __VAR__="__ATTR_VALUE__" label="__LABEL__"/>
```
```suggestion
         <param __ATTR__="__ATTR_VALUE__" label="__LABEL__"/>
```
